### DISCUSSION TITLE
Tidy up of VGG10 RadioML model

### DISCUSSION THREAD

**Discussion by satishkumar538** [Xilinx/finn#420] [Q&A] [answered]
"the ONNX model has been tidied up by removing the input quantization (See the last paragraph of attached screenshot of readme file provided with finn-example of vgg10-radionl)" How I can perform this removal of input quantization on the onnx file of the sandbox repository https://github.com/Xilinx/brevitas-radioml-challenge-21 . Please suggest steps/ codes.
![Satish_Finn1](https://user-images.githubusercontent.com/36513144/141241639-554c8737-12a8-4f46-a7c9-4c4070739da1.png)
_

**Comment by fpjentzsch** [comment, marked as answer]
Hi,

I attached some Python code in a .txt file: [network_surgery.txt](https://github.com/Xilinx/finn/files/7520226/network_surgery.txt)

We used this code to perform the necessary "tidy up" (aka "network surgery") to make the model ready for FINN, but I have not tested it on the latest Brevitas/FINN versions. It is not very elegant because it works on the .onnx model and not on the Pytorch model level, but it gets the job done.

As you see, we remove all nodes before the first "Mul" node. In our case, this includes the (MultiThreshold->Add) sequence that represents the input quantization. We also replace the final Softmax/LogSoftmax node with a TopK node. The rest of it is mostly cleanup steps, some of which might be redundant or not needed anymore.

**Reply by Aakasha01Agarwal** [reply]
I have used the network_surgery.txt file. I have used the .onnx file provided in the sandbox directory ( https://github.com/Xilinx/brevitas-radioml-challenge-21  ) but I am getting the following error. I don't seem to understand how to resolve it. Kindly help. I am also attaching the netron visualization of the .onnx model.
![1](https://user-images.githubusercontent.com/58568036/143425814-a0aba367-191d-4771-addf-b1e1b725459c.png)
![ff](https://user-images.githubusercontent.com/58568036/143425832-80fb18f8-35c6-4680-9887-004743bc3abc.png)

**Reply by fpjentzsch** [reply]
You cannot use the provided "pretrained_VGG10_w8a8_20_export.onnx" directly in FINN for 2 reasons:
1) It is in a different kind of ONNX format (referred to as Quantized-ONNX or QONNX for short) which represents quantization information as separate "Quant" nodes. I realize this is confusing, but the current FINN version still relies on a special kind of ONNX that Brevitas can export specifically for use in FINN. The attached code includes this export using the `FINNManager.export()` call.
2) The activation & weight bit width of 8 is likely too much to implement an efficient accelerator that fits the design, at least with the folding configuration given in finn-examples. To start your own experiments, I would try to set all bit widths to 4 and go from there.

**Reply by Aakasha01Agarwal** [reply]
Thank you so much for the quick response. I incorporated your suggestions and finally now I am able to run the network_surgery.txt correctly. I used the sandbox model and my own model and both of them are working correctly on your network_surgey.txt file. 

But after the network surgery I am supposed to use this model_ready.onnx for hardware deployment {code screenshot attached}. When I use the sandbox model the build is working fine but when I use my own model (screenshot attached) I am getting the following error.

I thank you again for your co-operation, kindly help.
# MY OWN MODEL
![7](https://user-images.githubusercontent.com/58568036/143443714-ef5b0ed2-0710-409c-92f0-0a07fb021fb5.png)
# CODE
![3](https://user-images.githubusercontent.com/58568036/143443808-4ac05f77-a57e-4f74-b28c-fd0aace0f838.png)
![4](https://user-images.githubusercontent.com/58568036/143443815-66378d95-9c95-48fa-8b2c-796c0989f929.png)
![5](https://user-images.githubusercontent.com/58568036/143443826-db108e2e-c833-49de-b1e0-d00561bfe106.png)
![6](https://user-images.githubusercontent.com/58568036/143443833-128bffe9-0389-45c4-aa1f-196d13218492.png)

**Reply by fpjentzsch** [reply]
This is likely caused by your AvgPool1d layer. FINN requires 1d layers to be converted into 2d layers (with one dimension set to 1) first (see https://github.com/Xilinx/finn/discussions/418#discussioncomment-1618707 for more background). This is the transformation that performs this conversion:
https://github.com/Xilinx/finn-base/blob/dev/src/finn/transformation/change_3d_tensors_to_4d.py

As you see, it only handles MaxPool right now. Since AveragePool operates almost the same way, it should be no problem to add support for it. You would only need to add "AveragePool" to the [list of valid nodes](https://github.com/Xilinx/finn-base/blob/dev/src/finn/transformation/change_3d_tensors_to_4d.py#L50) and [here](https://github.com/Xilinx/finn-base/blob/dev/src/finn/transformation/change_3d_tensors_to_4d.py#L189).

For a quick fix, you should be able to use an `nn.AvgPool2d` layer with one kernel dimension set to 1.

**Reply by Aakasha01Agarwal** [reply]
Thank you so much for your response. I have modified my model and now I am using Maxpool1d in place of Avgpool1d layer. I have successfully done the network_surgery on this modified model. When I am using this modified model for hardware deployment (codes: #420 (reply in thread)) I am getting errors (See Screenshot).

Please help.
# Model
![7](https://user-images.githubusercontent.com/58568036/143854232-d10ee8c9-8749-4692-86ab-e074f617e046.png)

# Errors
![8](https://user-images.githubusercontent.com/58568036/143854254-35704ef1-576f-4b2e-a73e-8ab1fbca27e8.png)

**Reply by fpjentzsch** [reply]
Could you show me the ONNX model after the network surgery and after the "convert_to_hls" step?

I see that your model does not include a `nn.LogSoftmax(dim=1)` layer at the end (you probably do this within the loss function). The code I provided assumes existence of this node, so maybe that's where things go wrong?

**Reply by Aakasha01Agarwal** [reply]
Thanks for your reply. See my ONNX model after network surgery and after the "convert_to_hls" step. Still I am not able to resolve the errors.
Please suggest.
# ONNX After Network_Surgery
![After_Network_Surgery](https://user-images.githubusercontent.com/58568036/144224685-40b5e879-bb12-4dd3-9df6-1c5a0ee5cb6b.png)

# ONNX after convert_to_hls
![After_hls](https://user-images.githubusercontent.com/58568036/144224751-117c4498-81e0-4eb1-b3e5-e90af74fcb3e.png)

**Reply by fpjentzsch** [reply]
I see, the problem is the feature map dimension across your model, which prevents the "MaxPoolNHWC" layers from being converted to "StreamingMaxPool_Batch" HLS layers.

[Here](https://github.com/Xilinx/finn/blob/30aa49f35adf89d244cd87c6050abbbfa161a75b/src/finn/transformation/fpgadataflow/convert_to_hls_layers.py#L355) you see that the kernel dimension (2 or 32 in your case) must divide the input dimension evenly. You would need to add additional padding or change your conv layers, so that you no longer have odd dimensions (e.g. 1025, 513, 257, ...) going into the MaxPool.

**Reply by Aakasha01Agarwal** [reply]
Dear fpjentzsch
Thanks for your continuous support. In my model, I have modified the kernel dimension and now I have even dimensions only. After running I get some other errors (see screenshot)
# Error
![9](https://user-images.githubusercontent.com/58568036/144238339-11ecd371-2b98-43cf-bf7d-b719a92118c2.png)
)
# ONNX after Network_Surgery
![Surgery_Model](https://user-images.githubusercontent.com/58568036/144240544-8addbd27-6f0e-4575-b61c-661628d51ecd.png)

**Reply by fpjentzsch** [reply]
This means that the folding_config_file you set in the build_cfg defines folding factors (PE & SIMD) that are incompatible with the dimensions of a layer (e.g. if you have 16 output channels you cannot parallelize PE>16). You can:
a) Adjust the folding factors for each layer in this file, e.g. set PE to divide the MH=16 for the offending layer in this example.
b) Do not define a `folding_config_file` in your build_cfg, but set a `target_fps` value. FINN will then try to automatically determine folding factors. In some cases though, even the automatic variant will produce illegal folding factors. A new version is on its way: https://github.com/Xilinx/finn/pull/438

**Comment by Aakasha01Agarwal** [comment]
Hello @fpjentzsch , as directed by you [here](https://github.com/Xilinx/finn/discussions/420#discussioncomment-1700568), I added "AveragePool" to the list of [valid nodes](https://github.com/Xilinx/finn-base/blob/dev/src/finn/transformation/change_3d_tensors_to_4d.py#L50) and [here ](https://github.com/Xilinx/finn-base/blob/dev/src/finn/transformation/change_3d_tensors_to_4d.py#L189). The network surgery operations works fine. But when I observe the .onnx file of the model, after the surgery, I can see that an extra layer named "PAD" is generated before the "AveragePool" layer. We should note that whenever we use a "MaxPool" layer, no such "PAD" layer is generated before it.

Using this model for hardware deployment I am getting the following error which I am pretty sure is caused by this extra "PAD" layer only.
![gg](https://user-images.githubusercontent.com/58568036/144602798-9b360cb0-7d06-40a5-a0f4-35e3f4079537.png)
![log](https://user-images.githubusercontent.com/58568036/144602998-3aca9288-c415-4ce1-b96a-df76e9a784c5.png)

I am attaching the error screenshot and the log file screenshot too.

**Reply by fpjentzsch** [reply]
Weird, if you see the "Pad" node this early in the flow (before streamlining), that means it must be inserted during the Brevitas export. I'm not sure why the export would create such a node. Since it does nothing, maybe you can remove it manually during network surgery with a few lines of code?

**Reply by Aakasha01Agarwal** [reply]
Thanks for the suggestion. I removed the "Pad" node successfully. Now after running the hardware deployment file I am still getting an error (screenshot attached).
![ttt](https://user-images.githubusercontent.com/58568036/144702681-1b4763a0-b6c1-453c-93b9-95de36f76f5a.png)

I am also attaching the .onnx file after the "convert_to_hls" step. I found that the "AveragePool" layer is not getting converted like "MaxPool" is getting converted to "StreamingMaxPool_Batch" HLS layer. I think this is caused because "AveragePool" layer is not defined in FINN library.
![tt](https://user-images.githubusercontent.com/58568036/144702685-dcaf026f-25d0-462a-99c3-b937424985ba.png)

Please suggest as to how to resolve this issue or let me know which FINN file needs to be updated for incorporating this AveragePool Layer.

**Reply by fpjentzsch** [reply]
Oh, I never worked with AvgPool myself, so I missed the following:
AveragePool is supported via the [Pool_Batch](https://github.com/Xilinx/finn/blob/30aa49f35adf89d244cd87c6050abbbfa161a75b/src/finn/custom_op/fpgadataflow/pool_batch.py#L37) HLSCustomOp and the [InferPool_Batch()](https://github.com/Xilinx/finn/blob/30aa49f35adf89d244cd87c6050abbbfa161a75b/src/finn/transformation/fpgadataflow/convert_to_hls_layers.py#L379) transformation instead of InferStreamingMaxPool(). Also, it only seems to work with the special Brevitas layer [QuantAvgPool2d](https://github.com/Xilinx/brevitas/blob/affa9cf10fa66bb84298cd3ad4e70fc576b6a9d3/src/brevitas/nn/quant_avg_pool.py#L57) and not `nn.AvgPool1d`. Pool_Batch was recently updated (https://github.com/Xilinx/finn/commit/123892ffc884ff55a5bf9ff1d6f385fd24d87a86) to support non-square inputs (you could set one dimension of QuantAvgPool2d to 1), but I'm not sure if it actually works in the current dev branch. Might be worth a try.
